File: repos/openrappter/typescript/src/memory/manager.ts
/**
 * Memory manager - manages chunks, embeddings, and search
 */

import type {
  MemoryChunk,
  MemorySearchResult,
  MemoryStatus,
  SearchOptions,
  MemorySource,
  EmbeddingProvider,
  MemoryManagerOptions,
} from './types.js';
import { chunkContent, generateSnippet, hashContent } from './chunker.js';

let chunkIdCounter = 0;

export class MemoryManager {
  private chunks: Map<string, MemoryChunk> = new Map();
  private embeddings: Map<string, number[]> = new Map();
  private embeddingCache: Map<string, number[]> = new Map();
  private provider: EmbeddingProvider | null;
  private chunkSize: number;
  private chunkOverlap: number;
  private pendingSync = 0;
  private lastSync?: string;

  constructor(options: MemoryManagerOptions = {}) {
    this.provider = options.embeddingProvider ?? null;
    this.chunkSize = options.chunkSize ?? 512;
    this.chunkOverlap = options.chunkOverlap ?? 50;
  }

  async add(content: string, source: MemorySource, sourcePath: string = '', metadata?: Record<string, unknown>): Promise<string> {
    const chunks = chunkContent(content, {
      chunkSize: this.chunkSize,
      overlap: this.chunkOverlap,
    });

    const ids: string[] = [];

    for (const chunkText of chunks) {
      const id = `mem_${Date.now()}_${++chunkIdCounter}`;
      const chunk: MemoryChunk = {
        id,
        content: chunkText,
        source,
        sourcePath,
        metadata,
        createdAt: new Date().toISOString(),
      };

      this.chunks.set(id, chunk);
      ids.push(id);

      // Generate embedding if provider available
      if (this.provider) {
        const cacheKey = hashContent(chunkText);
        let embedding = this.embeddingCache.get(cacheKey);

        if (!embedding) {
          const [result] = await this.provider.embed([chunkText]);
          embedding = result;
          this.embeddingCache.set(cacheKey, embedding);
        }

        chunk.embedding = embedding;
        this.embeddings.set(id, embedding);
      }
    }

    return ids[0];
  }

  async search(query: string, options: SearchOptions = {}): Promise<MemorySearchResult[]> {
    const limit = options.limit ?? 10;
    const threshold = options.threshold ?? 0;
    const sources = options.sources;

    let results: MemorySearchResult[] = [];

    // Vector search if provider available
    if (this.provider && this.embeddings.size > 0) {
      const [queryEmbedding] = await this.provider.embed([query]);
      results = this.vectorSearch(queryEmbedding, query, sources);
    }

    // FTS fallback/supplement
    const ftsResults = this.ftsSearch(query, sources);

    // Merge results (hybrid search)
    const merged = this.mergeResults(results, ftsResults);

    // Filter by threshold and limit
    return merged
      .filter(r => r.score >= threshold)
      .slice(0, limit);
  }

  async searchFts(query: string, options: SearchOptions = {}): Promise<MemorySearchResult[]> {
    const limit = options.limit ?? 10;
    const sources = options.sources;

    return this.ftsSearch(query, sources).slice(0, limit);
  }

  remove(id: string): boolean {
    const existed = this.chunks.delete(id);
    this.embeddings.delete(id);
    return existed;
  }

  removeBySourcePath(sourcePath: string): number {
    let removed = 0;
    for (const [id, chunk] of this.chunks) {
      if (chunk.sourcePath === sourcePath) {
        this.chunks.delete(id);
        this.embeddings.delete(id);
        removed++;
      }
    }
    return removed;
  }

  getStatus(): MemoryStatus {
    return {
      totalChunks: this.chunks.size,
      indexedChunks: this.embeddings.size,
      pendingSync: this.pendingSync,
      lastSync: this.lastSync,
    };
  }

  getChunk(id: string): MemoryChunk | undefined {
    return this.chunks.get(id);
  }

  listChunks(source?: MemorySource): MemoryChunk[] {
    const all = Array.from(this.chunks.values());
    if (source) {
      return all.filter(c => c.source === source);
    }
    return all;
  }

  clear(): void {
    this.chunks.clear();
    this.embeddings.clear();
  }

  private vectorSearch(queryEmbedding: number[], query: string, sources?: MemorySource[]): MemorySearchResult[] {
    const results: MemorySearchResult[] = [];

    for (const [id, embedding] of this.embeddings) {
      const chunk = this.chunks.get(id);
      if (!chunk) continue;
      if (sources && !sources.includes(chunk.source)) continue;

      const score = this.cosineSimilarity(queryEmbedding, embedding);
      results.push({
        chunk,
        score,
        snippet: generateSnippet(chunk.content, query),
      });
    }

    return results.sort((a, b) => b.score - a.score);
  }

  private ftsSearch(query: string, sources?: MemorySource[]): MemorySearchResult[] {
    const queryTerms = query.toLowerCase().split(/\s+/).filter(t => t.length > 2);
    if (queryTerms.length === 0) return [];

    const results: MemorySearchResult[] = [];

    for (const chunk of this.chunks.values()) {
      if (sources && !sources.includes(chunk.source)) continue;

      const contentLower = chunk.content.toLowerCase();
      let matchCount = 0;

      for (const term of queryTerms) {
        if (contentLower.includes(term)) {
          matchCount++;
        }
      }

      if (matchCount > 0) {
        const score = matchCount / queryTerms.length;
        results.push({
          chunk,
          score,
          snippet: generateSnippet(chunk.content, query),
        });
      }
    }

    return results.sort((a, b) => b.score - a.score);
  }

  private mergeResults(vectorResults: MemorySearchResult[], ftsResults: MemorySearchResult[]): MemorySearchResult[] {
    const merged = new Map<string, MemorySearchResult>();
    const vectorWeight = 0.7;
    const ftsWeight = 0.3;

    for (const r of vectorResults) {
      merged.set(r.chunk.id, { ...r, score: r.score * vectorWeight });
    }

    for (const r of ftsResults) {
      const existing = merged.get(r.chunk.id);
      if (existing) {
        existing.score += r.score * ftsWeight;
      } else {
        merged.set(r.chunk.id, { ...r, score: r.score * ftsWeight });
      }
    }

    return Array.from(merged.values()).sort((a, b) => b.score - a.score);
  }

  private cosineSimilarity(a: number[], b: number[]): number {
    if (a.length !== b.length) return 0;

    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    for (let i = 0; i < a.length; i++) {
      dotProduct += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }

    const denom = Math.sqrt(normA) * Math.sqrt(normB);
    if (denom === 0) return 0;

    return dotProduct / denom;
  }
}

File: repos/openrappter/python/openrappter/memory/manager.py
import time
import uuid
from .chunker import chunk_content, generate_snippet


class MemoryChunk:
    def __init__(self, chunk_id, content, source='', source_path='', metadata=None):
        self.id = chunk_id
        self.content = content
        self.source = source
        self.source_path = source_path
        self.metadata = metadata or {}
        self.created_at = time.time()


class MemoryManager:
    def __init__(self, chunk_size=512, chunk_overlap=50):
        self._chunks = {}  # id -> MemoryChunk
        self._chunk_size = chunk_size
        self._chunk_overlap = chunk_overlap
        self._counter = 0

    def _next_id(self):
        self._counter += 1
        return f'mem_{int(time.time())}_{self._counter}'

    def add(self, content, source='', source_path='', metadata=None):
        """Add content, chunk it, store chunks. Return first chunk id."""
        chunks = chunk_content(content, self._chunk_size, self._chunk_overlap)
        if not chunks:
            return None
        first_id = None
        for chunk_text in chunks:
            chunk_id = self._next_id()
            self._chunks[chunk_id] = MemoryChunk(chunk_id, chunk_text, source, source_path, metadata)
            if first_id is None:
                first_id = chunk_id
        return first_id

    def search_fts(self, query, sources=None):
        """Full-text search across chunks."""
        terms = [t.lower() for t in query.split() if len(t) > 2]
        if not terms:
            return []

        results = []
        for chunk in self._chunks.values():
            if sources and chunk.source not in sources:
                continue
            lower = chunk.content.lower()
            matches = sum(1 for t in terms if t in lower)
            if matches > 0:
                score = matches / len(terms)
                results.append({
                    'score': score,
                    'chunk': {
                        'id': chunk.id,
                        'content': chunk.content,
                        'source': chunk.source,
                    },
                })

        results.sort(key=lambda r: r['score'], reverse=True)
        return results

    def get_status(self):
        return {'total_chunks': len(self._chunks)}

    def get_chunk(self, chunk_id):
        chunk = self._chunks.get(chunk_id)
        if not chunk:
            return None
        return {'id': chunk.id, 'content': chunk.content, 'source': chunk.source}

    def list_chunks(self, source=None):
        chunks = self._chunks.values()
        if source:
            chunks = [c for c in chunks if c.source == source]
        return [{'id': c.id, 'content': c.content, 'source': c.source} for c in chunks]

    def remove(self, chunk_id):
        return self._chunks.pop(chunk_id, None) is not None

    def clear(self):
        self._chunks.clear()
